File: frontend/src/shared/domain/recurring.ts
/**
 * Pure, framework-free recurrence math — the TS mirror of
 * `backend/app/shared/domain/recurring.py`. Both cores must agree on the same
 * dates and the same deterministic occurrence ids (the mirrored tests assert
 * identical literals).
 *
 * Dates are handled as ISO `YYYY-MM-DD` strings (never `Date` objects) to avoid
 * timezone drift — recurrence is a calendar concept, not an instant.
 */

export type Frequency = "daily" | "weekly" | "biweekly" | "monthly" | "quarterly" | "yearly";

const MONTHS_STEP: Partial<Record<Frequency, number>> = {
  monthly: 1,
  quarterly: 3,
  yearly: 12,
};

interface Ymd {
  y: number;
  m: number; // 1-12
  d: number; // 1-31
}

function parse(iso: string): Ymd {
  const [y, m, d] = iso.split("-").map(Number);
  return { y, m, d };
}

function format({ y, m, d }: Ymd): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${y}-${pad(m)}-${pad(d)}`;
}

function daysInMonth(y: number, m: number): number {
  return new Date(Date.UTC(y, m, 0)).getUTCDate();
}

function addDays(iso: string, days: number): string {
  const { y, m, d } = parse(iso);
  const dt = new Date(Date.UTC(y, m - 1, d + days));
  return format({ y: dt.getUTCFullYear(), m: dt.getUTCMonth() + 1, d: dt.getUTCDate() });
}

function addMonths(iso: string, months: number): string {
  const { y, m, d } = parse(iso);
  const total = m - 1 + months;
  const year = y + Math.floor(total / 12);
  const month = ((total % 12) + 12) % 12 + 1;
  const day = Math.min(d, daysInMonth(year, month)); // clamp (Jan 31 → Feb 28)
  return format({ y: year, m: month, d: day });
}

export function advance(iso: string, frequency: Frequency, interval = 1): string {
  const step = interval < 1 ? 1 : interval;
  switch (frequency) {
    case "daily":
      return addDays(iso, step);
    case "weekly":
      return addDays(iso, 7 * step);
    case "biweekly":
      return addDays(iso, 14 * step);
    default: {
      const months = MONTHS_STEP[frequency];
      if (months === undefined) throw new Error(`Unknown frequency: ${frequency}`);
      return addMonths(iso, months * step);
    }
  }
}

export function occurrencesBetween(
  start: string,
  frequency: Frequency,
  interval: number,
  rangeStart: string,
  rangeEnd: string,
  endDate?: string | null,
): string[] {
  const out: string[] = [];
  let current = start;
  let guard = 0;
  while (current <= rangeEnd && guard < 10_000) {
    if (endDate && current > endDate) break;
    if (current >= rangeStart) out.push(current);
    current = advance(current, frequency, interval);
    guard += 1;
  }
  return out;
}

// --- Deterministic occurrence id (FNV-1a, mirrors the Python core) ------
const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const MASK64 = (1n << 64n) - 1n;

function fnv1a64(data: string): bigint {
  let h = FNV_OFFSET;
  const bytes = new TextEncoder().encode(data);
  for (const byte of bytes) {
    h = ((h ^ BigInt(byte)) * FNV_PRIME) & MASK64;
  }
  return h;
}

/** Deterministic UUID-format id for a rule's occurrence on `isoDate`. */
export function occurrenceId(ruleId: string, isoDate: string): string {
  const base = `${ruleId}:${isoDate}`;
  const hi = fnv1a64(base);
  const lo = fnv1a64(`${base}#saldo`);
  const hex = (hi.toString(16).padStart(16, "0") + lo.toString(16).padStart(16, "0")).slice(0, 32);
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20, 32)}`;
}

File: backend/app/shared/domain/recurring.py
"""Pure, framework-free recurrence math for recurring rules / bills.

No I/O and no framework imports (mirrors `shared/domain/budgeting`). The TS core
in `frontend/src/shared/domain/recurring.ts` must agree on the same results —
the mirrored tests assert the same dates and the same occurrence ids.

Two responsibilities:
  1. Advancing a date by a frequency (with monthly/yearly day-clamping).
  2. Deriving a *deterministic* occurrence id from (rule_id, date), so the same
     bill occurrence materialized on two devices collapses to one transaction
     under last-write-wins sync instead of duplicating.

The occurrence id is a 128-bit value rendered in UUID format. It is computed
with FNV-1a (not stdlib `uuid5`) precisely because the identical algorithm has
to run in TypeScript, where `uuid5`/SHA-1 is not readily available synchronously.
"""

from __future__ import annotations

import uuid
from calendar import monthrange
from datetime import date, timedelta
from typing import Literal

Frequency = Literal["daily", "weekly", "biweekly", "monthly", "quarterly", "yearly"]

_MONTHS_STEP = {"monthly": 1, "quarterly": 3, "yearly": 12}


def _add_months(d: date, months: int) -> date:
    total = d.month - 1 + months
    year = d.year + total // 12
    month = total % 12 + 1
    day = min(d.day, monthrange(year, month)[1])  # clamp (e.g. Jan 31 → Feb 28)
    return date(year, month, day)


def advance(d: date, frequency: Frequency, interval: int = 1) -> date:
    """Return the next occurrence date after `d` for the given frequency."""
    if interval < 1:
        interval = 1
    if frequency == "daily":
        return d + timedelta(days=interval)
    if frequency == "weekly":
        return d + timedelta(weeks=interval)
    if frequency == "biweekly":
        return d + timedelta(weeks=2 * interval)
    step = _MONTHS_STEP.get(frequency)
    if step is not None:
        return _add_months(d, step * interval)
    raise ValueError(f"Unknown frequency: {frequency}")


def occurrences_between(
    start: date,
    frequency: Frequency,
    interval: int,
    range_start: date,
    range_end: date,
    end_date: date | None = None,
) -> list[date]:
    """All occurrence dates in [range_start, range_end], honouring `end_date`.

    Iterates from `start`, advancing by the frequency, collecting dates that fall
    within the window. Bounded so a misconfigured rule can't loop forever.
    """
    out: list[date] = []
    current = start
    guard = 0
    while current <= range_end and guard < 10_000:
        if end_date is not None and current > end_date:
            break
        if current >= range_start:
            out.append(current)
        current = advance(current, frequency, interval)
        guard += 1
    return out


# --- Deterministic occurrence id ---------------------------------------
_FNV_OFFSET = 0xCBF29CE484222325
_FNV_PRIME = 0x100000001B3
_MASK64 = (1 << 64) - 1


def _fnv1a_64(data: bytes) -> int:
    h = _FNV_OFFSET
    for byte in data:
        h = ((h ^ byte) * _FNV_PRIME) & _MASK64
    return h


def occurrence_id(rule_id: str | uuid.UUID, d: date) -> uuid.UUID:
    """Deterministic UUID for the occurrence of `rule_id` on date `d`.

    Two independent materializations of the same bill occurrence therefore
    produce the *same* transaction id, so sync dedupes them instead of creating
    duplicate transactions.
    """
    base = f"{rule_id}:{d.isoformat()}"
    hi = _fnv1a_64(base.encode("utf-8"))
    lo = _fnv1a_64((base + "#saldo").encode("utf-8"))
    value = (hi << 64) | lo
    return uuid.UUID(int=value)
